README
This is a "plugin" for the Video Disk Recorder (VDR) to access DVB-streams
produced by the NetCeiver-hardware.

Project's homepage:
http://www.baycom.de/hardware/netceiver/

Latest version available via GIT at:
https://github.com/vdr-projects/vdr-plugin-mcli

The vdr-plugin is
Copyright (C) 2007-2010 by BayCom GmbH <info@baycom.de>

Extensions by contributors
Copyright (C) 2020-2023 by Peter Bieringer <pb@bieringer.de>
Copyright (C) 2013-2015 by hsteinhaus
Copyright (C) 2014      by mhorwath
Copyright (C) 2013      by Hoppaz


It is dynamically linked against "libnetceiver", available via GIT at:
https://github.com/vdr-projects/libnetceiver

"netcv2dvbip" was carved out into a dedicated project available via GIT at:
https://github.com/vdr-projects/netcv2dvbip

------------------------------------------------------------------

The vdr-plugin is covered by the
GNU GENERAL PUBLIC LICENSE (GPL), version 2:

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 2 as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

**** Reviewed version ****

- Support at least vdr >= 2.4.4

== Potential requirements ==

On newer Linux systems allow vdr to open raw sockets:

setcap cap_net_raw=pe /usr/sbin/vdr


for systemd:

cat <<END >/etc/systemd/system/vdr.service.d/vdr-mcli.conf
[Service]
AmbientCapabilities=CAP_NET_RAW
END


== Plugin options ==

--ifname <interface>
	network interface of the multicast client
		the interface where the Netceiver boxes are also connected

--port <port>
	used network port (default: 23000)

--dvb-s <num>
--dvb-c <num>
--dvb-t <num>
--atsc <num>
--dvb-s2 <num>
	limit number of devices (default: 8)

--mld-reporter-disable
	disable mld reporter

--sock-path <path>
	path of socket directory, default: API_SOCK_NAMESPACE

--cam-disable (NEW since 0.9.5)
	reject tuning to channels which requiring CAM

--tuner-max <num> (NEW since 0.9.5)
	limit maximum number of tuners to <num>
		avoid that LCARS skin shows 8 tuners in a 3 tuner system

--debugmask <mask> (NEW since 0.9.5, see logging.h)
	<mask> can be hexadecimal (0x..) or decimal
	conditionally enable debug messages
		PIDS          0x01
		TUNE_EXTRA    0x02
		TUNE          0x04
		RESOURCES     0x08
		PIDS_ADD_DEL  0x10
		TUNE_PC       0x40    // ProvideChannel
		FILTER        0x80

--logskipmask <mask> (NEW since 0.9.6, see logging.h)
	<mask> can be hexadecimal (0x..) or decimal
	conditionally disable log messages
		recv_ts_func/Data PIDs	0x01    // skip log of issues with Data pids (0 or 16-18) like Mcli::recv_ts_func: Discontinuity on receiver 0x559f735c7e00 for pid 18: 5->7 at pos 0/7
		SetChannelDevice/Reject	0x02    // skip log of Mcli::SetChannelDevice: Reject tuning on DVB (interesting if cam-disable is active)

--netcvupdate-use-lftp
	add "lftp" specific option to "netcvupdate" calls

--netcvupdate-enable-debug
	add debug option to "netcvupdate" calls


== RPM Build Support included ==

# download vdr-devel, rpm build environment and required dependencies:
dnf install vdr-devel rpm-build gcc-c++ libxml2-devel zlib-devel

# specify version:
VERSION=1.0.0

# download version with proper tar.gz name:
wget https://github.com/vdr-projects/vdr-plugin-mcli/archive/v$VERSION/vdr-plugin-mcli-$VERSION.tar.gz

# run rpm rebuild with tar.gz:
rpmbuild -tb vdr-plugin-mcli-$VERSION.tar.gz

# install (example, here 1.0.0 on Fedora 38)
dnf localinstall $HOME/rpmbuild/RPMS/x86_64/vdr-mcli-1.0.0-5.fc38.x86_64.rpm

== Packages ==

* RPM (Fedora): until included in upstream available via COPR: https://copr.fedorainfracloud.org/coprs/pbiering/vdr_extensions/packages/


HISTORY
VDR Plugin 'mcli' Revision History
----------------------------------
2023-06-05 1.0.1
- pbiering move call of InitMcli from cPluginMcli::Start to cPluginMcli::Initialize (https://github.com/vdr-projects/vdr-plugin-mcli/pull/17)

2023-06-01 1.0.0
- pbiering add DeviceName, contribution from https://github.com/vdr-projects/vdr-plugin-mcli/issues/15

2023-05-31
- M-Reimer adjustments to link dynamically against "libnetceiver"
- pbiering adjust spec file

2022-02-07 0.9.7
- pbiering fix lftp support for netcvupdate and cam_menu.c
- pbiering add new options for plugin: --netcvupdate-use-lftp --netcvupdate-enable-debug

2022-01-22
- pbiering fix some log statements and REELVDR related defines

2021-06-12 0.9.6
- pbiering several log extensions

2021-04-03 0.9.6pre8
- pbiering fix idle priority handling

2021-03-30
- pbiering add additional option '-n' to netcvupdate to use 'lftp' as FTP client

2021-03-21 0.9.6pre7
- extend with (currently hardcoded timings) interface watcher

2021-03-16 0.9.6pre6
- pbiering apply adjusted fixes from https://www.vdr-portal.de/forum/index.php?thread/134181-mcli-error-video-data-stream-broken/&postID=1337872#post1337872

2021-03-16 0.9.6pre5
- pbiering add error/debug log for PreInit related to network device

2021-03-13 0.9.6pre4
- pbiering in case of --cam-disable option is active, do not enable triggerCam

2021-03-12 0.9.6pre3
- pbiering apply vdr-plugin-mcli-0.9.6n_05-FixSetChannelDeviceCamAlloc.patch from https://www.vdr-portal.de/forum/index.php?thread/134181-mcli-error-video-data-stream-broken/&postID=1337625#post1337625

2021-03-12 0.9.6pre2
- pbiering apply vdr-plugin-mcli-0.9.6m_02-DeviceReadyWithCi.patch from https://www.vdr-portal.de/forum/index.php?thread/134181-mcli-error-video-data-stream-broken/&postID=1337593#post1337593
- pbiering apply vdr-plugin-mcli-0.9.6m_01-TriggerCamAllPids.patch from https://www.vdr-portal.de/forum/index.php?thread/134181-mcli-error-video-data-stream-broken/&postID=1337593#post1337593

2021-03-11 0.9.6pre1
- pbiering introduce PreInit and PostExit to avoid crashes in case of other plugins have fatal issues and VDR has to stop instantly

2021-02-25
- pbiering add triggerCam patch from https://www.vdr-portal.de/forum/index.php?thread/134181-mcli-error-video-data-stream-broken/&postID=1336731#post1336731

2021-02-08
- pbiering fix spec file for rpmbuild -ta

2021-02-07
- pbiering add more header include fixes based on contribution from HelmutB

2021-02-06
- pbiering add additional special debug masks (see README), align/fix log prefix to "mcli::"

2021-02-05
- pbiering add log skip mask option (--logskipmask) (see logging.h)
- pbiering improve debugging enhancement based on contribution from HelmutB

2021-02-02
- pbiering include not covered patches from https://www.minidvblinux.de/forum/index.php?topic=9526.0

2021-01-30
- pbiering add vdr/systemd/override.conf into RPM (via file in contrib subdirectory)

2021-01-26: Version 0.9.5
- pbiering add contrib directory with RPM spec and vdr-plugin config file template

2021-01-25
- pbiering cMcliDevice::SetPid: LOCK_THREAD deactivated because resulting in deadlock with osdteletext and zapping
- pbiering honor cam-disable earlier in cMcliDevice::SetChannelDevice

2021-01-24
- pbiering add debug mask option (--debugmask) (see mcli.h)
- pbiering fix a bunch of debug messages and turn them to be controllable by --debug-mask
- pbiering add cam-disable mode which avoid tuning to channel which requires CAM in case CAMs are generally not in use in the system (--cam-disable)

2020-12-15
- pbiering add support for DeviceType (supported since 1.7.28)
- pbiering add tuner-max option to limit amount of tuners in skins
- pbiering implement rate limiter for "Discontinuity on receiver" messages
- implement missing GetCurrentlyTunedTransponder (supported since 1.7.13)

2019-11-24
- pbiering fix several compiler warnings and incompatibilities to vdr-2.4.1 (0.9.3 work-in-progress)

2015-11-12
- hsteinhaus debugged makefiles to ensure complete compilation, even when using parallel make jobs
- hsteinhaus consolidated debug logging and disabled it by default
- hsteinhaus check if CAM is present before enforcing it's use
- hsteinhaus resolved some compiler warnings
- hsteinhaus removed some unused headers

2014-07-12
- mhorwath debug 2.1.x issue http://projects.vdr-developer.org/issues/1709 , little message cleanup, added cam info at startup, fixed compile warning  

2013-09-20
- Hoppaz applied svn diff -r188:189 from https://svn.baycom.de/repos/vdr-mcli-plugin/ : r189 | fliegl | 2013-09-19 21:26:55 +0200 (Do, 19. Sep 2013) | 2 Zeilen , Add LGPL license for all tools (including netcvdvbip)
 
2013-09-18
- hsteinhaus explicitely define API_SOCK to resolve compile problems using Debian/Ubuntu tools (reason why OS-depended define does not work is not fully understood)
- hsteinhaus explicitely set -fPIC for the plugin to resolve Debian build problems
- Hoppaz fixed makefile netcv2dvbip du to link error
- Hoppaz added netcv2dvbip revision 188 from https://svn.baycom.de/repos/vdr-mcli-plugin/mcast/netcv2dvbip
- hsteinhaus Resolved issue #1504 - compile problems related to overloading of const methods of the cDevice interface
- Hoppaz removed obsolete .svn directories
- Hoppaz patch svn-pull.patch.txt applied
- Hoppaz patch new_make_format.patch.txt applied
- Hoppaz patch signal_strength.patch.txt applied
 
2010-06-23:
- Moved the mcli-library to LGPL-2.1-license.

2008-12-18: Version 0.0.1
- Initial revision.


logging.h
/* debug option */
#define DEBUG_PIDS
#define DEBUG_TUNE_EXTRA
#define DEBUG_TUNE
#define DEBUG_RESOURCES
#define DEBUG_FILTER
#define DEBUG_PIDS_ADD_DEL
#define DEBUG_TUNE_PC

#define DEBUG_BIT_PIDS	 	0x01
#define DEBUG_BIT_TUNE_EXTRA	0x02
#define DEBUG_BIT_TUNE		0x04
#define DEBUG_BIT_RESOURCES	0x08
#define DEBUG_BIT_PIDS_ADD_DEL 	0x10
#define DEBUG_BIT_TUNE_PC	0x40	// ProvideChannel
#define DEBUG_BIT_FILTER	0x80
#define DEBUG_BIT_SIGNAL	0x100	// mcli::handle_ten

// hidden test options
#define DEBUG_BIT_Action_RetuneOnFirstTuner    0x1000  // retune if the first tuner is found (cPluginMcli::Action)
#define DEBUG_BIT_recv_ts_func_NO_LOGRATELIMIT 0x2000  // disable rate limiter Mcli::recv_ts_func
#define DEBUG_BIT_Action_SkipTriggerCam        0x4000  // skip trigger CAM initialization, even if the first tuning is for a FTA program (cMcliDevice)

#define DEBUG_MASK(bit, code)	if ((m_debugmask & bit) != 0) { code };

extern int m_debugmask;


/* Log skip option */

#define LOGSKIP_BIT_recv_ts_func_pid_Data	0x01	// skip log of issues with Data pids (16-18) like Mcli::recv_ts_func: Discontinuity on receiver 0x559f735c7e00 for pid 18: 5->7 at pos 0/7
#define LOGSKIP_BIT_SetChannelDevice_Reject	0x02	// skip log of Mcli::SetChannelDevice: Reject tuning on DVB

#define LOGSKIP_MASK(bit, code)	if ((m_logskipmask & bit) == 0) { code };

extern int m_logskipmask;
